File: arkhe-monorepo/catedral_os_v152/python/hardware/undulator.py
# python/hardware/undulator.py
"""
Nó Undulator — decaimento físico com Zeno Veto (I194.I197) e adaptativo (I198).
Mirror Python do crate Rust `arkhe-catedral-v152`.

  I194  Zeno Veto: dt <= 2 * ranging
  I195  Ranging Delta (microssegundos da linha)
  I196  Decaimento exponencial: Phi(t) = Phi0 * exp(-lambda * t)
  I198  lambda(t) = lambda_0 + beta * (1 - f_handover), clamp em [min,max]
"""

from __future__ import annotations

import math
import struct
import time
from dataclasses import dataclass
from typing import List, Optional

from core.spectral import irreducible_coherence


@dataclass(frozen=True)
class RangingDelta:
    """I195: ranging da linha em microssegundos."""
    us: int

    def as_secs(self) -> float:
        return self.us * 1e-6

    def zeno_threshold(self) -> int:
        """I194: limiar de Zeno = 2 * ranging."""
        return self.us * 2


@dataclass(frozen=True)
class DecayRate:
    """I196: taxa de decaimento (1/tau)."""
    value: float

    @classmethod
    def from_ranging(cls, ranging: RangingDelta) -> "DecayRate":
        tau = ranging.as_secs()
        return cls(1.0 / tau if tau > 0.0 else 1.0)

    def apply(self, phi: float, dt_secs: float) -> float:
        return phi * math.exp(-self.value * dt_secs)


@dataclass
class UndulatorHandover:
    phi: float
    source: str
    timestamp: float
    ranging_delta_us: int
    zeno_passed: bool
    decayed_phi: float


@dataclass
class AdaptiveParams:
    """I198: parametros de adaptacao do decaimento."""
    initial_decay: float = 0.001
    adaptation_rate: float = 0.0005
    min_decay: float = 0.0001
    max_decay: float = 0.01


class UndulatorNode:
    """Nó Undulator com Zeno Veto (I194) e decaimento adaptativo (I198)."""

    def __init__(self, ranging: RangingDelta,
                 adaptive: Optional[AdaptiveParams] = None):
        self.ranging = ranging
        self.adaptive = adaptive or AdaptiveParams()
        self.decay_rate = DecayRate.from_ranging(ranging) if adaptive is None \
            else DecayRate(self.adaptive.initial_decay)
        self._last_handover_ts = time.monotonic()
        self._handover_times: List[float] = []
        self._count = 0
        self._phi_received = 0.0
        self._phi_accepted = 0.0

    # ------------------------------------------------------------------ #
    # Frequencia / adaptacao (I198)
    # ------------------------------------------------------------------ #
    def _update_decay_rate(self) -> None:
        if self.adaptive.adaptation_rate <= 0.0:
            return
        now = time.monotonic()
        window = 10.0  # segundos
        self._handover_times = [t for t in self._handover_times if t >= now - window]
        freq = len(self._handover_times) / window
        raw = (self.adaptive.initial_decay
               + self.adaptive.adaptation_rate * (1.0 - min(freq, 1.0)))
        self.decay_rate = DecayRate(
            max(self.adaptive.min_decay, min(self.adaptive.max_decay, raw)))

    # ------------------------------------------------------------------ #
    # Handover (I194)
    # ------------------------------------------------------------------ #
    def receive_handover(self, phi: float, source: str = "Z1T") -> Optional[UndulatorHandover]:
        now = time.monotonic()
        dt = now - self._last_handover_ts
        dt_us = int(dt * 1e6)
        dt_secs = dt

        zeno_ok = dt_us <= self.ranging.zeno_threshold()
        if not zeno_ok:
            return None

        self._handover_times.append(now)
        self._update_decay_rate()

        decayed = self.decay_rate.apply(phi, dt_secs)
        self._last_handover_ts = now
        self._count += 1
        self._phi_received += phi
        self._phi_accepted += decayed

        return UndulatorHandover(
            phi=phi,
            source=source,
            timestamp=time.time(),
            ranging_delta_us=self.ranging.us,
            zeno_passed=True,
            decayed_phi=decayed,
        )

    # ------------------------------------------------------------------ #
    # Estadisticas
    # ------------------------------------------------------------------ #
    def stats(self) -> dict:
        return {
            "handover_count": self._count,
            "total_phi_received": self._phi_received,
            "total_phi_accepted": self._phi_accepted,
            "acceptance_rate": (self._phi_accepted / self._phi_received
                                if self._phi_received > 0 else 0.0),
            "ranging_us": self.ranging.us,
            "decay_rate": self.decay_rate.value,
        }

    def encode_packet(self, phi: float, source: int = 0) -> bytes:
        """Empacota um handover para o barramento UART (payload G1)."""
        ts_us = int(time.time() * 1e6)
        return struct.pack("<Bfq", source & 0xFF, phi, ts_us)

    @staticmethod
    def phi_reference() -> float:
        """I157 aplicado ao nó: referencia quasi-universal (N=13)."""
        return irreducible_coherence(13)

File: arkhe-monorepo/packages/arkhe-catedral-v152/src/hardware/undulator.rs
//! Nó Undulator — decaimento físico com Zeno Veto (I194–I197) e adaptativo
//! (I198).

use crate::core::spectral::irreducible_coherence;
use serde::{Deserialize, Serialize};
use std::time::{Instant, SystemTime, UNIX_EPOCH};

/// I195 — Ranging Delta (microssegundos de linha).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct RangingDelta(pub u64);

impl RangingDelta {
    pub fn from_micros(us: u64) -> Self {
        Self(us)
    }

    pub fn as_micros(&self) -> u64 {
        self.0
    }

    pub fn as_secs_f64(&self) -> f64 {
        self.0 as f64 * 1e-6
    }

    /// I194 — Limiar de Zeno = 2 × ranging.
    pub fn zeno_threshold(&self) -> u64 {
        self.0 * 2
    }
}

/// I196/I198 — Taxa de decaimento.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct DecayRate(pub f64);

impl DecayRate {
    pub fn from_ranging(ranging: RangingDelta) -> Self {
        let tau = ranging.as_secs_f64();
        if tau <= 0.0 {
            Self(1.0)
        } else {
            Self(1.0 / tau)
        }
    }

    /// Decaimento exponencial: Φ(t) = Φ₀·exp(−λ·t).
    pub fn apply(&self, phi: f64, dt_secs: f64) -> f64 {
        phi * (-self.0 * dt_secs).exp()
    }

    pub fn value(&self) -> f64 {
        self.0
    }
}

/// Um handover aceito pelo Undulator.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UndulatorHandover {
    pub phi: f64,
    pub source: String,
    pub timestamp: u64,
    pub ranging_delta_us: u64,
    pub zeno_passed: bool,
    pub decayed_phi: f64,
}

/// Estatísticas agregadas do nó Undulator.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct UndulatorStats {
    pub handover_count: u64,
    pub total_phi_received: f64,
    pub total_phi_accepted: f64,
    pub acceptance_rate: f64,
    pub ranging_us: u64,
    pub decay_rate: f64,
}

/// Nó Undulator com Zeno Veto (I194) e decaimento adaptativo (I198).
pub struct UndulatorNode {
    pub ranging_delta: RangingDelta,
    pub decay_rate: DecayRate,
    last_handover: Instant,
    handover_count: u64,
    total_phi_received: f64,
    total_phi_accepted: f64,
    history: Vec<UndulatorHandover>,
    /// I198 — parâmetros de adaptação.
    adaptive: AdaptiveParams,
}

/// I198 — λ(t) = λ₀ + β·(1 − f), com clamp em [min, max].
#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct AdaptiveParams {
    pub initial_decay: f64,
    pub adaptation_rate: f64,
    pub min_decay: f64,
    pub max_decay: f64,
}

impl Default for AdaptiveParams {
    fn default() -> Self {
        Self {
            initial_decay: 0.001,
            adaptation_rate: 0.0005,
            min_decay: 0.0001,
            max_decay: 0.01,
        }
    }
}

impl UndulatorNode {
    pub fn new(ranging_delta: RangingDelta) -> Self {
        let decay_rate = DecayRate::from_ranging(ranging_delta);
        Self {
            ranging_delta,
            decay_rate,
            last_handover: Instant::now(),
            handover_count: 0,
            total_phi_received: 0.0,
            total_phi_accepted: 0.0,
            history: Vec::new(),
            adaptive: AdaptiveParams::default(),
        }
    }

    pub fn with_adaptive(mut self, ap: AdaptiveParams) -> Self {
        self.adaptive = ap;
        self.decay_rate = DecayRate(ap.initial_decay);
        self
    }

    /// I194 — Recebe handover com verificação do limiar de Zeno.
    pub fn receive_handover(&mut self, phi: f64, source: &str) -> Option<UndulatorHandover> {
        let now = Instant::now();
        let dt = now - self.last_handover;
        let dt_us = dt.as_micros() as u64;
        let dt_secs = dt.as_secs_f64();
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        self.handover_count += 1;
        self.total_phi_received += phi;

        let zeno_threshold = self.ranging_delta.zeno_threshold();
        let zeno_passed = dt_us <= zeno_threshold;

        if !zeno_passed {
            return None;
        }

        // I198 — taxa adaptativa baseada na frequência de handovers.
        if self.adaptive.adaptation_rate > 0.0 {
            self.decay_rate = DecayRate(self.adaptive_decay(now));
        }

        let decayed_phi = self.decay_rate.apply(phi, dt_secs);

        let handover = UndulatorHandover {
            phi,
            source: source.to_string(),
            timestamp,
            ranging_delta_us: self.ranging_delta.as_micros(),
            zeno_passed: true,
            decayed_phi,
        };

        self.history.push(handover.clone());
        self.last_handover = now;
        self.total_phi_accepted += decayed_phi;

        Some(handover)
    }

    /// I198 — λ(t) = λ₀ + β·(1 − f_handover), clampada.
    fn adaptive_decay(&self, now: Instant) -> f64 {
        let ap = self.adaptive;
        let freq = self.frequency(now);
        let raw = ap.initial_decay + ap.adaptation_rate * (1.0 - freq.min(1.0));
        raw.clamp(ap.min_decay, ap.max_decay)
    }

    /// Frequência estimada de handovers por segundo (janela fixa de 10s).
    fn frequency(&self, now: Instant) -> f64 {
        let window = std::time::Duration::from_secs_f64(10.0);
        let cutoff = now - window;
        // A janela de timestamps reais exige armazenamento; aproximação
        // determinística: todos os handovers aceitos caem na janela se o
        // nó recebeu `count` desde o início.
        let _ = cutoff;
        self.history.len() as f64 / 10.0
    }

    pub fn stats(&self) -> UndulatorStats {
        UndulatorStats {
            handover_count: self.handover_count,
            total_phi_received: self.total_phi_received,
            total_phi_accepted: self.total_phi_accepted,
            acceptance_rate: if self.total_phi_received > 0.0 {
                self.total_phi_accepted / self.total_phi_received
            } else {
                0.0
            },
            ranging_us: self.ranging_delta.as_micros(),
            decay_rate: self.decay_rate.value(),
        }
    }

    pub fn accepted_phi_reference(&self) -> f64 {
        // I157 aplicada ao nó: referência quasi-universal.
        irreducible_coherence(13)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_zeno_bound() {
        let ranging = RangingDelta(1000);
        let mut undulator = UndulatorNode::new(ranging);

        std::thread::sleep(std::time::Duration::from_micros(500));
        let result = undulator.receive_handover(0.95, "test");
        assert!(result.is_some());

        std::thread::sleep(std::time::Duration::from_micros(2500));
        let result = undulator.receive_handover(0.95, "test");
        assert!(result.is_none());
    }

    #[test]
    fn test_decay_application() {
        let ranging = RangingDelta(1000);
        let decay = DecayRate::from_ranging(ranging);
        let result = decay.apply(1.0, 0.001);
        assert!((result - 0.367).abs() < 0.01);
    }

    #[test]
    fn test_zeno_threshold() {
        assert_eq!(RangingDelta(1000).zeno_threshold(), 2000);
    }

    #[test]
    fn test_adaptive_decay_bounded() {
        let ranging = RangingDelta(1000);
        let mut u = UndulatorNode::new(ranging).with_adaptive(AdaptiveParams::default());
        let Some(h) = u.receive_handover(0.9, "src") else {
            panic!("should accept when fresh");
        };
        assert!(h.decayed_phi > 0.0);
        assert!(u.stats().decay_rate < u.adaptive.max_decay + 1e-12);
        assert!(u.stats().decay_rate > u.adaptive.min_decay - 1e-12);
    }

    #[test]
    fn test_stats() {
        let ranging = RangingDelta(1000);
        let mut u = UndulatorNode::new(ranging);
        std::thread::sleep(std::time::Duration::from_micros(200));
        u.receive_handover(0.8, "a");
        let s = u.stats();
        assert_eq!(s.handover_count, 1);
        assert!(s.acceptance_rate <= 1.0 + 1e-9);
    }
}